File: frontend/src/types/temporada.ts
import type { TipoGenero, TipoCategoria, TipoTorneo } from "../constants/enums"

/** Papel de un torneo dentro de su temporada anual. */
export type RolEnTemporada = "REGULAR" | "NO_COMPUTA" | "FINAL_ANUAL"

export interface TorneoEnTemporada {
  id_torneo: number
  nombre: string
  tipo: TipoTorneo
  rol_en_temporada: RolEnTemporada | null
  fecha_inicio: string | null
  fecha_fin: string | null
  activo: boolean
}

/** Agrupación anual de torneos de una misma liga (apertura + clausura). */
export interface Temporada {
  id_temporada: number
  nombre: string
  anio: number
  categoria: TipoCategoria
  division?: string | null
  genero: TipoGenero
  activa: boolean
  torneos: TorneoEnTemporada[]
  creado_en: string
  actualizado_en: string | null
  borrado_en: string | null
  creado_por: string | null
  actualizado_por: string | null
}

/** Fila de la tabla anual: la campaña de un equipo a lo largo del año. */
export interface FilaTablaAnual {
  id_temporada: number
  temporada: string
  anio: number
  categoria: TipoCategoria
  division?: string | null
  genero: TipoGenero

  id_equipo: number
  equipo: string
  id_club: number

  puesto: number
  /** De cuántos torneos REGULAR viene la campaña (para ver quién jugó medio año). */
  torneos_computados: number
  partidos_jugados: number
  ganados: number
  empatados: number
  perdidos: number
  goles_a_favor: number
  goles_en_contra: number
  diferencia_gol: number
  puntos: number
}

/** Torneo candidato del selector de la tabla anual, con su estado actual. */
export interface TorneoComputable {
  id_torneo: number
  nombre: string
  tipo: TipoTorneo
  fecha_inicio: string | null
  activo: boolean
  /** Si hoy suma a la tabla anual. */
  computa: boolean
}

/** Datos del selector: la liga, su temporada si ya existe, y los candidatos. */
export interface SelectorTablaAnual {
  anio: number
  categoria: TipoCategoria
  division?: string | null
  genero: TipoGenero
  id_temporada: number | null
  nombre_temporada: string | null
  torneos: TorneoComputable[]
}

/** Ronda con la que arranca el playoff anual; define cuántos clasifican. */
export type RondaInicialAnual = "octavos" | "cuartos" | "semifinal" | "final"

export const EQUIPOS_POR_RONDA_ANUAL: Record<RondaInicialAnual, number> = {
  octavos: 16,
  cuartos: 8,
  semifinal: 4,
  final: 2,
}

/** Duelo de la primera ronda cuando el bracket se arma a mano. */
export interface DueloAnual {
  id_equipo_local: number
  id_equipo_visitante: number
}

export interface CrearPlayoffAnualRequest {
  nombre: string
  fecha_inicio?: string | null
  ronda_inicial: RondaInicialAnual
  formato: "ida" | "ida_y_vuelta"
  asignacion: "automatico" | "manual"
  duelos?: DueloAnual[] | null
  tercer_puesto: boolean
  es_competitiva?: boolean
  /** Heredar al playoff la última nómina del año de cada clasificado. */
  copiar_planteles: boolean
}

/** Integrante que no se pudo heredar al playoff, con el motivo. */
export interface PlantelOmitido {
  equipo: string
  id_persona: number
  nombre: string
  motivo: string
}

export interface PlayoffAnualResponse {
  id_torneo: number
  nombre: string
  id_temporada: number
  /** Clasificados, en orden de mérito. */
  equipos: string[]
  planteles_copiados: number
  planteles_omitidos: PlantelOmitido[]
  /** Equipos que no tenían nómina propia en ningún torneo del año. */
  avisos: string[]
}

File: backend/app/schemas/temporada.py
# app/schemas/temporada.py
from datetime import datetime, date
from typing import Optional, List, Literal

from pydantic import BaseModel, Field, ConfigDict

from app.models.enums import CategoriaTipo, GeneroTipo, RolTorneoTemporada, TipoTorneo
from app.schemas.fixture_playoff import DueloManual


class TemporadaBase(BaseModel):
    nombre: str = Field(..., min_length=1, max_length=100)
    anio: int = Field(..., ge=1900, le=2200)
    categoria: CategoriaTipo
    division: Optional[str] = Field(None, max_length=30)
    genero: GeneroTipo
    activa: bool = True


class TemporadaCreate(TemporadaBase):
    pass


class TemporadaUpdate(BaseModel):
    nombre: Optional[str] = Field(None, min_length=1, max_length=100)
    activa: Optional[bool] = None


class TorneoEnTemporada(BaseModel):
    """Torneo asociado a una temporada, con el papel que juega en ella."""
    id_torneo: int
    nombre: str
    tipo: TipoTorneo
    rol_en_temporada: Optional[RolTorneoTemporada] = None
    fecha_inicio: Optional[date] = None
    fecha_fin: Optional[date] = None
    activo: bool = True

    model_config = ConfigDict(from_attributes=True)


class TemporadaSchema(TemporadaBase):
    id_temporada: int
    creado_en: datetime
    actualizado_en: Optional[datetime] = None
    borrado_en: Optional[datetime] = None
    creado_por: Optional[str] = None
    actualizado_por: Optional[str] = None

    torneos: List[TorneoEnTemporada] = []

    model_config = ConfigDict(from_attributes=True)


class AsignarTorneoRequest(BaseModel):
    """Asocia un torneo a la temporada con un rol, o lo desasocia.

    Con `rol = None` el torneo sale de la temporada. La restricción
    `chk_torneo_rol_requiere_temporada` impide dejar un rol sin temporada.
    """
    id_torneo: int = Field(..., gt=0)
    rol: Optional[RolTorneoTemporada] = None


class FilaTablaAnual(BaseModel):
    """Fila de la tabla anual (vista `vw_tabla_posiciones_anual`)."""
    id_temporada: int
    temporada: str
    anio: int
    categoria: str
    division: Optional[str] = None
    genero: str

    id_equipo: int
    equipo: str
    id_club: int

    puesto: int
    torneos_computados: int
    partidos_jugados: int
    ganados: int
    empatados: int
    perdidos: int
    goles_a_favor: int
    goles_en_contra: int
    diferencia_gol: int
    puntos: int

    model_config = ConfigDict(from_attributes=True)


class TorneoComputable(BaseModel):
    """Torneo candidato del selector de la tabla anual, con su estado actual."""
    id_torneo: int
    nombre: str
    tipo: TipoTorneo
    fecha_inicio: Optional[date] = None
    activo: bool = True
    computa: bool

    model_config = ConfigDict(from_attributes=True)


class SelectorTablaAnual(BaseModel):
    """Lo que necesita el selector: la liga, su temporada (si ya existe) y los
    torneos candidatos con el tilde puesto en los que suman."""
    anio: int
    categoria: CategoriaTipo
    division: Optional[str] = None
    genero: GeneroTipo
    id_temporada: Optional[int] = None
    nombre_temporada: Optional[str] = None
    torneos: List[TorneoComputable] = []


class DefinirComputablesRequest(BaseModel):
    """Estado final del selector: los torneos tildados para la tabla anual."""
    anio: int = Field(..., ge=1900, le=2200)
    categoria: CategoriaTipo
    division: Optional[str] = Field(None, max_length=30)
    genero: GeneroTipo
    id_torneos: List[int] = Field(default_factory=list)


# ── Playoff por el campeón anual ──────────────────────────────────────────────

RondaInicialAnual = Literal["octavos", "cuartos", "semifinal", "final"]


class CrearPlayoffAnualRequest(BaseModel):
    """Alta del playoff por el campeón del año de una temporada.

    A diferencia del playoff común, no se elige un torneo base: los equipos
    salen de la tabla anual de la temporada. La ronda inicial es obligatoria
    porque es la que define cuántos clasifican (octavos 16, cuartos 8,
    semifinal 4, final 2).

    En `automatico` clasifican los mejores de la tabla y se siembran mejor
    contra peor. En `manual` el admin arma los duelos de la primera ronda —
    solo con equipos de la tabla anual — y de ahí en adelante el bracket sigue
    solo con los ganadores.
    """
    nombre: str = Field(..., min_length=1, max_length=100)
    fecha_inicio: Optional[date] = None
    ronda_inicial: RondaInicialAnual
    formato: Literal["ida", "ida_y_vuelta"] = "ida"
    asignacion: Literal["automatico", "manual"] = "automatico"
    duelos: Optional[List[DueloManual]] = None
    tercer_puesto: bool = False
    es_competitiva: bool = True
    # Copiar al playoff la última nómina del año de cada equipo clasificado.
    copiar_planteles: bool = True


class PlantelOmitido(BaseModel):
    """Integrante que no se pudo heredar al playoff, con el motivo."""
    equipo: str
    id_persona: int
    nombre: str
    motivo: str


class PlayoffAnualResponse(BaseModel):
    """Resultado del alta: el torneo creado y qué quedó pendiente a mano."""
    id_torneo: int
    nombre: str
    id_temporada: int
    equipos: List[str] = []
    planteles_copiados: int = 0
    planteles_omitidos: List[PlantelOmitido] = []
    # Equipos que no tenían nómina propia en ningún torneo del año.
    avisos: List[str] = []
